discord/ext/commands/converter.py

```diff
@@ -159,6 +159,65 @@ class UserConverter(IDConverter):
 
         return result
 
+class MessageConverter(IDConverter):
+    """Converts to a :class:`discord.Message`.
+
+    The lookup strategy is as follows (in order):
+
+    1. Lookup by "{channel ID}-{message ID}" (got by shift-clicking on "Copy ID")
+    2. Lookup by message ID (the message **must** be in the context channel)
+    3. Lookup by message URL
+    """
+
+    async def convert(self, ctx, argument):
+        prog = re.compile(r'^(?:(?P<channel_id>[0-9]{15,21})-)?(?P<message_id>[0-9]{15,21})$')
+        match = prog.match(argument)
+        if match:
+            message_id = int(match.group("message_id"))
+            message = ctx.bot._connection._get_message(message_id)
+            if message:
+                return message
+            channel_id = match.group("channel_id")
+            if channel_id:
+                channel = ctx.bot.get_channel(int(channel_id))
+                if not channel:
+                    raise BadArgument('Channel "{channel}" not found.'.format(channel=channel_id))
+                try:
+                    message = await channel.fetch_message(message_id)
+                except discord.errors.NotFound:
+                    raise BadArgument('Message "{msg}" not found.'.format(msg=argument))
+                except discord.errors.Forbidden:
+                    raise BadArgument("Can't read messages in {channel}".format(channel=channel.mention))
+            else:
+                channel = ctx.channel
+                try:
+                    message = await channel.fetch_message(message_id)
+                except discord.errors.NotFound:
+                    raise BadArgument('Message "{msg}" not found.'.format(msg=argument))
+        else:
+            prog = re.compile(
+                r'^https?://(?:(ptb|canary)\.)?discordapp\.com/channels/'
+                r'(?P<guild_id>[0-9]{15,21})/(?P<channel_id>[0-9]{15,21})/(?P<message_id>[0-9]{15,21})$'
```

Comment:
This regex does not support `@me` in the URLs, which is for private message contexts.